cpp/src/io/parquet/experimental/hybrid_scan_helpers.cpp

```diff
@@ -0,0 +1,140 @@
+using row_group_info                 = parquet::detail::row_group_info;
+
+metadata::metadata(cudf::host_span<uint8_t const> footer_bytes)
+{
+  CompactProtocolReader cp(footer_bytes.data(), footer_bytes.size());
+  cp.read(this);
+  CUDF_EXPECTS(cp.InitSchema(this), "Cannot initialize schema");
+  sanitize_schema();
+}
+
+aggregate_reader_metadata::aggregate_reader_metadata(cudf::host_span<uint8_t const> footer_bytes,
+                                                     bool use_arrow_schema,
+                                                     bool has_cols_from_mismatched_srcs)
+  : aggregate_reader_metadata_base({}, false, false)
+{
+  // Re-initialize internal variables here as base class was initialized without a source
+  per_file_metadata = std::vector<metadata_base>{metadata{footer_bytes}.get_file_metadata()};
+  keyval_maps       = collect_keyval_metadata();
+  schema_idx_maps   = init_schema_idx_maps(has_cols_from_mismatched_srcs);
+  num_rows          = calc_num_rows();
+  num_row_groups    = calc_num_row_groups();
+
+  // Force all columns to be nullable
+  auto& schema = per_file_metadata.front().schema;
+  std::for_each(schema.begin(), schema.end(), [](auto& col) {
+    col.repetition_type = FieldRepetitionType::OPTIONAL;
+  });
+
+  // Collect and apply arrow:schema from Parquet's key value metadata section
+  if (use_arrow_schema) {
+    apply_arrow_schema();
+
+    // Erase ARROW_SCHEMA_KEY from the output pfm if exists
+    std::for_each(keyval_maps.begin(), keyval_maps.end(), [](auto& pfm) {
+      pfm.erase(cudf::io::parquet::detail::ARROW_SCHEMA_KEY);
+    });
+  }
+}
+
+cudf::io::text::byte_range_info aggregate_reader_metadata::get_page_index_bytes() const
+{
+  auto& schema     = per_file_metadata.front();
+  auto& row_groups = schema.row_groups;
+
+  if (row_groups.size() and row_groups.front().columns.size()) {
+    int64_t const min_offset = schema.row_groups.front().columns.front().column_index_offset;
+    auto const& last_col     = schema.row_groups.back().columns.back();
+    int64_t const max_offset = last_col.offset_index_offset + last_col.offset_index_length;
+    return {min_offset, (max_offset - min_offset)};
+  }
+
+  return {};
+}
+
+FileMetaData const& aggregate_reader_metadata::get_parquet_metadata() const
+{
+  return per_file_metadata.front();
+}
+
+void aggregate_reader_metadata::setup_page_index(cudf::host_span<uint8_t const> page_index_bytes)
+{
+  // Return early if empty page index buffer span
+  if (not page_index_bytes.size()) {
+    CUDF_LOG_WARN("Hybrid scan reader encountered empty `PageIndex` buffer");
+    return;
+  }
+
+  auto& schema     = per_file_metadata.front();
+  auto& row_groups = schema.row_groups;
+
+  CUDF_EXPECTS(row_groups.size() and row_groups.front().columns.size(),
+               "No column chunks in Parquet schema to read PageIndex for");
+
+  CompactProtocolReader cp(page_index_bytes.data(), page_index_bytes.size());
```

Comment:
Snippet copied over directly from `metadata::metadat()` in `reader_impl_helpers.cpp`